app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -234,91 +195,298 @@ async function addEthereumChainHandler(
+  end,
+  {
+    upsertNetworkConfiguration,
+    getCurrentRpcUrl,
+    findNetworkConfigurationBy,
+    setActiveNetwork,
+    requestUserApproval,
+    startApprovalFlow,
+    endApprovalFlow,
+    getCurrentChainIdForDomain,
+    getCaveat,
+    requestSwitchNetworkPermission,
+    getChainPermissionsFeatureFlag,
+  },
+) {
+  const validParams = validateParams(req.params[0], end);
+  if (!validParams) {
+    return end();
+  }
+
+  const {
+    chainId,
+    chainName,
+    firstValidBlockExplorerUrl,
+    firstValidRPCUrl,
+    ticker,
+  } = validParams;
+  const { origin } = req;
+
+  const currentChainIdForDomain = getCurrentChainIdForDomain(origin);
+  const currentNetworkConfiguration = findExistingNetwork(
+    currentChainIdForDomain,
+    findNetworkConfigurationBy,
+  );
+
+  // TODO should this be searching by rpcUrl instead of chainId?
+  // currently we aren't preventing adding the same rpcUrl multiple times
+  // since this just finds the first network with the same chainId
+  // the actual add is prevented downstream in upsertNetworkConfiguration
+  // but we should consider changing this to search by rpcUrl
+  const existingNetwork = findExistingNetwork(
+    chainId,
+    findNetworkConfigurationBy,
+  );
+
+  // if the chainId is the same as an existing network but the ticker is different we want to block this action
+  // as it is potentially malicious and confusing
+  if (
+    existingNetwork &&
+    existingNetwork.chainId === chainId &&
+    existingNetwork.ticker !== ticker
+  ) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `nativeCurrency.symbol does not match currency symbol for a network the user already has added with the same chainId. Received:\n${ticker}`,
+      }),
+    );
+  }
+
+  let networkConfigurationId;
+  let requestData;
+  let approvalFlowId;
+
+  // If the network doesn't already exist with the same rpcUrl, add it
+  if (!existingNetwork || existingNetwork.rpcUrl !== firstValidRPCUrl) {
+    ({ id: approvalFlowId } = await startApprovalFlow());
+
+    try {
+      await requestUserApproval({
+        origin,
+        type: ApprovalType.AddEthereumChain,
+        requestData: {
+          chainId,
+          rpcPrefs: { blockExplorerUrl: firstValidBlockExplorerUrl },
+          chainName,
+          rpcUrl: firstValidRPCUrl,
+          ticker,
+        },
+      });
+
+      networkConfigurationId = await upsertNetworkConfiguration(
+        {
+          chainId,
+          rpcPrefs: { blockExplorerUrl: firstValidBlockExplorerUrl },
+          nickname: chainName,
           rpcUrl: firstValidRPCUrl,
-          chainId: _chainId,
-          nickname: _chainName,
           ticker,
-          networkConfigurationId,
         },
-        fromNetworkConfiguration: getProviderConfig(),
+        { source: MetaMetricsNetworkEventSource.Dapp, referrer: origin },
+      );
+
+      // Once the network has been added, the request is considered successful
+      res.result = null;
```

Comment:
We ultimately call `switchChain()`, which can error, and should therefore let it handle setting the result or error properties. In other words, these lines should be deleted, and we should ensure that `switchChain()` always correctly ends the request.

Reply:
done here: https://github.com/MetaMask/metamask-extension/pull/24415/commits/29330d24b6c2116df638d228dab93245a375bd73